### DISCUSSION TITLE
ORM Foreign Keys & Joins

### DISCUSSION THREAD

**Discussion by elitru** [vlang/v#11868] [Ideas]
# ORM Foreign Keys & Joins

For SQL n:m relationships or 1:n relationships a foreign key join functionality would be useful.
## 1:n relationships
- One customer can have multiple orders

![grafik](https://user-images.githubusercontent.com/16387917/119327882-c06c6900-bc83-11eb-9d78-ac51f04931b3.png)
```v
struct Customer {
	customer_id int
	orders []Order // the order contains the user_id
}

struct Order {
     // customer Customer // some people might want to / need have a bidirectional connection
     ...
}
```
## n:m relationships | Case 1: No additional attributes inside the connection entity
- One customer can have multiple orders
- One order consits of one or more products

![grafik](https://user-images.githubusercontent.com/16387917/119328295-3d97de00-bc84-11eb-9612-3201e9c0dc41.png)

A useful model structure could look something like this:
```v
struct Customer {
	customer_id int
	orders []Order // the order contains the user_id
}

struct Order {
	products []Product
     // customer Customer // some people might want to / need have a bidirectional connection
}

struct Product {
	product_id int
	name string
}
```

## n:m relationships | Case 2: One additional attribute inside the connection entity
However, there is also another case where the models could look a little bit different, because of additional attributes inside the connection entities:
- One customer can have multiple orders
- One order consits of one or more products (however, you also need to know how much of each product you want to buy, so that's saved as a property inside the connection entity which in this case would be "ProductOrder")

![grafik](https://user-images.githubusercontent.com/16387917/119329465-8a2fe900-bc85-11eb-82cb-294dd044d03a.png)

In this case, the models could look something like this:
```v
struct Customer {
	customer_id int
	orders []Order
}

struct Order {
	products map[Product]int // each product is connected with its amount
    // customer Customer // some people might want to / need have a bidirectional connection
}

struct Product {
	product_id int
	name string
}
```

## n:m relationships: Case 3: More than one additional attribute inside the connection entity
If you need the save more than one additonal property inside the connection entity, you are likely to create a connection entity as an own model anyway. The connection entity would look the same as an 1:n relation.

![grafik](https://user-images.githubusercontent.com/16387917/119329588-ac296b80-bc85-11eb-8da6-33a01cf83b85.png)

```v
struct Customer {
	customer_id int
	orders []Order
}

struct Order {
	products map[Product]int // product connected with the amount
    // customer Customer // some people might want to / need have a bidirectional connection
}

struct Product {
	product_id int
	name string
}

struct ProductOrder {
	product Product
	order Order
	amount int
	additionalAttribute int
	...
}
```

**Comment by khalyomede** [comment]
Thank you a lot @elitru for such a detailed description, I like the way you expressed foreign keys using structs!

In order for the ORM to know to which column a field with a struct type belongs to, maybe we should give a hint to the dialect? I was thinking of something like this:

```v
struct ProductOrder {
  product Product [sql: foreign "product_id"]
  order Order [sql: foreign "order_id"]
  amount int
  additionalAttribute int
  ...
}
```

What do you think about it?

**Comment by elitru** [comment]
This sounds great! It's also easy to understand for anybody what the code actually does. :)
